### ISSUE TITLE
Declaration merging in plain JS with JSDoc

### ISSUE THREAD

**Issue by trusktr** [microsoft/TypeScript#57559]
### 🔍 Search Terms

jsdoc declaration merging

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

Ability to merge declarations. 

### 📃 Motivating Example

For example, this is a declaration merge in TS:

```ts
export const Foo = makeFooClass()
export type Foo = InstanceType<typeof Foo>
```

In JS we can't do this. I was hoping for something similar to this to achieve the same:

```js
export const Foo = makeFooClass() // returns a class
/** @export {InstanceType<typeof Foo>} Foo */
```

### 💻 Use Cases

1. What do you want to use this for? To merge declarations in certain cases.
2. What shortcomings exist with current approaches? There's no way to do it AFAICT
3. What workarounds are you using in the meantime? None

**Comment by fatcerberus** [comment]
I don’t think that counts as a declaration merge, it’s just “declaring a type and a value with the same name”.  There’s no requirement that the two `Foo`s be related as they exist in separate namespaces within the compiler.  Declaration merge is when e.g. you declare an interface more than once and the second one augments the first.

**Comment by trusktr** [comment]
Maybe its named something else. If so, what is this pattern named where a single identifier carries both a type and a value? I could not find anything about it in the TS docs (maybe I don't know what to search for).

**Comment by trusktr** [comment]
Maybe the syntax needs to be

```js
/**
 * @export
 * @typedef {InstanceType<typeof Foo>} Foo
 */
```

**Comment by trusktr** [comment]
Ah, @Gerrit0 pointed out that the following works because `typedef` is automatically exported:

```js
export const Foo = makeFooClass() // returns a class
/** @typedef {InstanceType<typeof Foo>} Foo */
```

**Comment by trusktr** [comment]
I do not see that this type exports feature is mentioned here:

https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html

Perhaps we need to add more info to that page.

**Comment by fatcerberus** [comment]
> If so, what is this pattern named where a single identifier carries both a type and a value? I could not find anything about it in the TS docs

I don't think there's a name for it because it's just declaring two different things that happen to have the same name, but the compiler considers them separate entities as one exists in type-space and the other in value-space[^1].  Nothing is really being merged.

[^1]: This split between type-space and value-space is why the `typeof` operator exists.  There's no requirement that `Foo` and `typeof Foo` be related, it's just in practice they usually are.

**Comment by trusktr** [comment]
> one exists in type-space and the other in value-space[1](#user-content-fn-1-541e3f148d502bab1fa97b7a48bb050e). Nothing is really being merged.

They're being "merged" because the end user who will import the type can reference it with a single identifier, and to them it is the "same thing".


```js
import {Foo} from './Foo.js'

const f = new Foo() // uses *both*, which to the user is a single thing.
```

**Comment by trusktr** [comment]
I stumbled on this again. I wish there were support for JSDoc declaration merging so that a project that likes plain JS doesn't need additional `.d.ts` files laying around.

How much is the bounty? :P

Another example based on a real project:

```js
export class Foo {...}
```
```js
// other file
import {Foo} from './Foo.js'

/** ...need some way to augment the `Foo` type definition with the new method... */
Foo.prototype.newMethod = function() {...}
```

**Comment by trusktr** [comment]
Or for example we need the equivalent of `declare foo: number` within a `class` definition in lieu of declaration merging for some cases.

EDIT One way to *kinda* `declare` in plain JS without defining a property is:

```js
class SomeClass {
  // this only works in TypeScript:
  // declare foo: number

  constructor() {
    // This works in JS:
    /** @type {number} */
    this.foo // access the declared property without actually setting it (Note! if it is a getter defined elsewhere, it will cause a `get` to happen)
  }
}
```

Te caveat is that, if `this.foo` actually exists, a `get` operation will happen (f.e. a getter will be triggered, with potential side-effects). In cases there the property doesn't exist beforehand, the `this.foo` access is benign.
